castai/resource_service_account.go

```diff
@@ -0,0 +1,250 @@
+			},
+			FieldServiceAccountDescription: {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Description of the service account.",
+			},
+			FieldServiceAccountEmail: {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Email of the service account.",
+			},
+			FieldServiceAccountAuthor: {
+				Type: schema.TypeList,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						FieldServiceAccountAuthorID:    {Type: schema.TypeString, Computed: true},
+						FieldServiceAccountAuthorEmail: {Type: schema.TypeString, Computed: true},
+						FieldServiceAccountAuthorKind:  {Type: schema.TypeString, Computed: true},
+					},
+				},
+				Computed:    true,
+				Description: "Author of the service account.",
+			},
+		},
+	}
+}
+
+func resourceServiceAccountRead(ctx context.Context, data *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*ProviderConfig).api
+
+	if data.Id() == "" {
+		return diag.Errorf("service account ID is not set")
+	}
+
+	organizationID := data.Get(FieldServiceAccountOrganizationID).(string)
+
+	tflog.Info(ctx, "reading service account", map[string]interface{}{
+		"resource_id":     data.Id(),
+		"organization_id": organizationID,
+	})
+
+	resp, err := client.ServiceAccountsAPIGetServiceAccountWithResponse(ctx, organizationID, data.Id())
+	if resp.StatusCode() == http.StatusNotFound {
+		tflog.Warn(ctx, "resource is not found, removing from state", map[string]interface{}{
+			"resource_id":     data.Id(),
+			"organization_id": organizationID,
+		})
+		data.SetId("") // Mark resource as deleted
+		return nil
+	}
+	if err := sdk.CheckOKResponse(resp, err); err != nil {
+		return diag.Errorf("getting service account: %v", err)
+	}
+
+	tflog.Info(ctx, "found service account", map[string]interface{}{
+		"resource_id":     data.Id(),
+		"organization_id": organizationID,
+	})
+	serviceAccount := resp.JSON200
+
+	if err := data.Set(FieldServiceAccountName, serviceAccount.ServiceAccount.Name); err != nil {
+		return diag.Errorf("setting service account name: %v", err)
+	}
+
+	if err := data.Set(FieldServiceAccountEmail, serviceAccount.ServiceAccount.Email); err != nil {
+		return diag.Errorf("setting service account email: %v", err)
+	}
+
+	if err := data.Set(FieldServiceAccountDescription, serviceAccount.ServiceAccount.Description); err != nil {
+		return diag.Errorf("setting service account description: %v", err)
+	}
+
+	authorData := flattenServiceAccountAuthor(serviceAccount.ServiceAccount.Author)
+	if err := data.Set(FieldServiceAccountAuthor, authorData); err != nil {
+		return diag.Errorf("setting service account author: %v", err)
+	}
+	return nil
+}
+
+func resourceServiceAccountCreate(ctx context.Context, data *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*ProviderConfig).api
+
+	organizationID := data.Get(FieldServiceAccountOrganizationID).(string)
```

Comment:
There is a way to get the default organization from the token, it can be used when the organization is empty:
```
organizationID, err = getDefaultOrganizationId(ctx, meta)
```

Reply:
updated